text/0112-compress-state-response-message-in-state-sync.md

```diff
@@ -0,0 +1,70 @@
+
+## Summary
+
+This RFC proposes compressing the state response message during the state syncing process to reduce the amount of data transferred.
+
+## Motivation
+
+State syncing can require downloading several gigabytes of data, particularly for blockchains with large state sizes, such as Astar, which
+has a state size exceeding 5 GiB (https://github.com/AstarNetwork/Astar/issues/1110). This presents a significant
+challenge for nodes with slower network connections. Additionally, the current state sync implementation lacks a persistence feature (https://github.com/paritytech/polkadot-sdk/issues/4),
+meaning any network disruption forces the node to re-download the entire state, making the process even more difficult.
+
+## Stakeholders
+
+This RFC benefits all projects utilizing the Substrate framework, specifically in improving the efficiency of state syncing.
+
+- Node Operators.
+- Substrate Users.
+
+## Explanation
+
+The largest portion of the state response message consists of either `CompactProof` or `Vec<KeyValueStateEntry>`, depending on whether a proof is requested ([source](https://github.com/paritytech/polkadot-sdk/blob/0cd577ba1c4995500eb3ed10330d93402177a53b/substrate/client/network/sync/src/state_request_handler.rs#L216-L241)):
+
+- `CompactProof`: When proof is requested, compression yields a lower ratio but remains beneficial, as shown in warp sync tests in the Performance section below.
+- `Vec<KeyValueStateEntry>`: Without proof, this is theoretically compressible because the entries are generated by iterating the
+storage sequentially starting from an empty storage key, which means many entries in the message share the same storage prefix, making it ideal
+for compression.
+
+## Drawbacks
+
+None identified.
+
+## Testing, Security, and Privacy
+
+The [code changes](https://github.com/liuchengxu/polkadot-sdk/commit/2556fefacd2e817111d838af5f46d3dfa495852d) required for this RFC are straightforward: compress the state response on the sender side and decompress it on the receiver side. Existing sync tests should ensure functionality remains intact.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+This RFC optimizes network bandwidth usage during state syncing, particularly for blockchains with gigabyte-sized states, while introducing negligible CPU overhead for compression and decompression. For example, compressing the state response during a recent Polkadot warp sync (around height #22076653) reduces the data transferred from 530,310,121 bytes to 352,583,455 bytes — a 33% reduction, saving approximately 169 MiB of data.
+
+Performance data is based on [this patch](https://github.com/liuchengxu/polkadot-sdk/commit/da93360c9a59c29409061789c598d8f4e55d7856), with logs available [here](https://github.com/liuchengxu/polkadot-sdk/commit/9d98cefd5fac0a001d5910f7870ead05ab99eeba).
+
+### Ergonomics
+
+None.
+
+### Compatibility
```

Comment:
It looks to me that this needs some network protocol versioning. For example what happens if the receiver is not expecting compressed data? Or if the receiver is expecting compressed data but gets uncompressed data?

Reply:
This is a valid concern, and I agree that network versioning would be the ideal solution. However, it seems that Substrate doesn't currently support network protocol versioning, which might deserve another RFC.

In the meantime, we can introduce a backward-compatible workaround. For example, we could define a new `StateRequestV1` with an additional `compress_response: bool` field. This field would default to `false` if not provided. Here's how it would work in practice, assuming an old node (OldNode) and an upgraded node (NewNode):

- OldNode to NewNode (OldNode => NewNode):
    When OldNode sends a `StateRequest` payload to NewNode, the NewNode interprets it as `StateRequestV1 { ..., compress_response: false }`. Based on this field, NewNode will decide whether to compress the response, ensuring that OldNode receives the data in an uncompressed format, maintaining compatibility.

- NewNode to OldNode (NewNode => OldNode):
    If NewNode sends a `StateRequestV1` payload to OldNode, OldNode will fail to decode it, leading to an aborted sync. This is because OldNode does not recognize the new `StateRequestV1` structure.

- OldNode to OldNode and NewNode to NewNode:
    Syncing between nodes of the same version (OldNode => OldNode or NewNode => NewNode) will proceed as expected, with OldNodes using uncompressed data and NewNodes using compressed data when possible.
    
The main consequence of this approach is that NewNodes won't be able to sync state from OldNodes, but they can still sync with other NewNodes. This workaround isn't perfect, but it ensures some level of compatibility until a more comprehensive network versioning solution can be implemented in Substrate.

Reply:
I added some support for req/response protocol versioning in https://github.com/paritytech/polkadot-sdk/pull/2771 and used it for the implementation of https://github.com/polkadot-fellows/RFCs/pull/47. Although it's not optimal. It would first try a request on the new version and if that's not supported (so making 2 round trips if the remote does not support the new version). It was implemented this way mostly due to libp2p not having the most flexible interfaces.

In any case, the RFC should clearly state the interactions between new/old nodes 

CC: @lexnv

Reply:
@bkchr Any more comments apart from the interactions between the old and new nodes? If the workaround I proposed in my last comment is fine, I can update the RFC.

Reply:
This optimization makes sense from my perspective. We are trading off bandwidth for CPU cycles in hopes that this will yield an overall faster warp sync. 

It would be interesting to see the comparison with times as well.
For example the [sub-triage-logs](https://github.com/lexnv/sub-triage-logs/) tool has a newly added command that measures the `warp-sync` time:

```
# Provide the path to the logs outputted by a substrate-based node
cargo run -- warp-time --file ../substrate-log.txt
```

Indeed we'd need to think about versioning as well. We'd like to deprecate the older request-response protocols, so its worth adding as few breaking changes as possible and having a deprecation plan in mind:
- when the NewNode sends out the V2 format to the OldNode, the old node will fail to decode. After a decoding error the OldNode will change the reputation of the NewNode and possibly disconnect the NewNode from its network
- NewNode can try to send out the V1 format after the V2 was dropped, but ideally, we should avoid the first case

Offhand, it looks like a similar approach to https://github.com/paritytech/polkadot-sdk/pull/2771 should solve this issue. Then if we are to introduce a new request-response version, should we also consider other compression algorithms? I think this complicates things, but would be interesting to see some comparison between different compression algorithms and polkadot specific data (ie maybe we end up saving 20% bandwidth instead of 33% but we are faster to compress)